app/gui/view/graph-editor/src/builtin/visualization/native/lazy_text_visualization/text_provider.rs

```diff
@@ -0,0 +1,269 @@
+    pub fn new(
+        receive_data: frp::Source<visualization::Data>,
+        preprocessor_update: frp::Any<PreprocessorConfiguration>,
+    ) -> Self {
+        let frp = text_provider::Frp::new();
+        let network = frp.network();
+        let output = frp.private().output.clone_ref();
+
+        frp::extend! { network
+            grid_window <- any_mut::<GridWindow>();
+        }
+
+        let grid_cache_update = f!([grid_window](new_window) {
+           grid_window.emit(new_window)
+        });
+
+        let text_cache = Rc::new(RefCell::new(GridCache::<String>::new(
+            GridPosition::new(0, 0),
+            GridSize::new(20, 20),
+            CACHE_PADDING,
+            Box::new(grid_cache_update),
+        )));
+
+        frp::extend! { network
+
+            register_access <- any_mut();
+
+            update_preprocessor <- all(register_access, grid_window);
+            update_preprocessor <- update_preprocessor._1();
+            preprocessor_update <+ update_preprocessor.map(|grid_window| {
+                let grid_posititon = grid_window.position.map(|value| value.max(0));
+                let grid_size = grid_window.size;
+                lazy_text_preprocessor(grid_posititon, grid_size)
+            });
+
+            grid_data_update <- receive_data.map(|data| {
+                LazyGridData::try_from(data.clone()).ok()
+            });
+            grid_data_update <- grid_data_update.map(|data| data.clone()).unwrap();
+            chunks_update <- grid_data_update.map(|grid_data| grid_data.chunks.clone());
+            eval chunks_update([text_cache](chunks) {
+                for (pos, text) in chunks {
+                    text_cache.borrow_mut().add_item(*pos, text.clone());
+                }
+            });
+
+            output.line_count <+ grid_data_update.map(|grid_data| grid_data.line_count);
+            output.longest_line <+ grid_data_update.map(|grid_data| grid_data.longest_line);
+
+        }
+
+        Self { frp, text_cache, register_access }
+    }
+}
+
+impl TextProvider for BackendTextProvider {
+    fn get_slice(&self, line: usize, chunk_index: usize) -> Option<String> {
+        let y = line as i32;
+        let x = chunk_index as i32;
+        let result = self.text_cache.borrow_mut().get_item(Vector2::new(x, y));
+        self.register_access.emit(());
+        result
+    }
+
+    fn frp(&self) -> &Frp {
+        &self.frp
+    }
+}
+
+/// Crate a preprocessor configuration for the lazy text preprocessor.
+fn lazy_text_preprocessor(
+    grid_position: GridPosition,
+    grids_size: GridSize,
+) -> PreprocessorConfiguration {
+    PreprocessorConfiguration::new(
+        "Standard.Visualization.Preprocessor",
+        "lazy_preprocessor",
+        vec![
+            serde_json::to_string(&grid_position).unwrap(),
+            serde_json::to_string(&grids_size).unwrap(),
+            serde_json::to_string(&(CHARS_PER_CELL as u32)).unwrap(),
```

Comment:
Are we OK with these `unwraps` ? When they can fail?

Reply:
I replaced them with excepts, but they cannot happen, as the values we serialize are always serializable as string.